java/ql/src/experimental/Security/CWE/CWE-927/SensitiveBroadcast.ql

```diff
@@ -0,0 +1,172 @@
+/**
+ * @name Broadcasting sensitive data to all Android applications
+ * @id java/sensitive-broadcast
+ * @description An Android application uses implicit intents to broadcast sensitive data to all applications without specifying any receiver permission.
+ * @kind path-problem
+ * @tags security
+ *       external/cwe-927
+ */
+
+import java
+import semmle.code.java.frameworks.android.Intent
+import semmle.code.java.dataflow.TaintTracking
+
+/**
+ * Gets a regular expression for matching common names of variables that indicate the value being held contains sensitive information.
+ */
+private string getCommonSensitiveInfoRegex() {
+  result = "(?i).*challenge|pass(wd|word|code|phrase)(?!.*question).*" or
+  result = "(?i).*(token|username|userid|secret).*"
+}
+
+/**
+ * Gets regular expression for matching names of Android variables that indicate the value being held contains sensitive information.
+ */
+private string getAndroidSensitiveInfoRegex() { result = "(?i).*(email|phone|ticket).*" }
+
+/**
+ * Method call to pass information to the `Intent` object.
+ */
+class PutIntentExtraMethodAccess extends MethodAccess {
+  PutIntentExtraMethodAccess() {
+    getMethod().getName().regexpMatch("put\\w*Extra(s?)") and
+    getMethod().getDeclaringType() instanceof TypeIntent
+  }
+}
+
+/**
+ * Method call to pass information to the intent extra bundle object.
+ */
+class PutBundleExtraMethodAccess extends MethodAccess {
+  PutBundleExtraMethodAccess() {
+    getMethod().getName().regexpMatch("put\\w+") and
+    getMethod().getDeclaringType().getASupertype*().hasQualifiedName("android.os", "BaseBundle")
+  }
+}
+
+/** Finds variables that hold sensitive information judging by their names. */
+class SensitiveInfoExpr extends Expr {
+  SensitiveInfoExpr() {
+    exists(Variable v | this = v.getAnAccess() |
+      (
+        v.getName().toLowerCase().regexpMatch(getCommonSensitiveInfoRegex()) or
+        v.getName().toLowerCase().regexpMatch(getAndroidSensitiveInfoRegex())
+      )
+    )
+  }
+}
+
+/**
+ * The method access of `context.sendBroadcast`.
+ */
+class SendBroadcastMethodAccess extends MethodAccess {
+  SendBroadcastMethodAccess() {
+    this.getMethod().getDeclaringType() instanceof TypeContext and
+    this.getMethod().getName().matches("send%Broadcast%")
+  }
+}
+
+private class NullArgFlowConfig extends DataFlow2::Configuration {
+  NullArgFlowConfig() { this = "Flow configuration with a null argument" }
+
+  override predicate isSource(DataFlow::Node src) { src.asExpr() instanceof NullLiteral }
+
+  override predicate isSink(DataFlow::Node sink) { any() }
+}
+
+private class EmptyArrayArgFlowConfig extends DataFlow2::Configuration {
```

Comment:
Use DataFlow3 -- it's technically fine to use the same version of dataflow when the flows are independent, like null pointers and empty arrays, but apparently it helps the compiler by telling it that the two configs definitely never interact.

Reply:
Changed <code>EmptyArrayArgFlowConfig</code> to use DataFlow3